File: dp_labs_and_solutions/dp_labs_and_solutions/dp_cpp_labs_and_solutions/composite/lab/multiple_journey.cpp
// multiple_journey.cpp

#include <iostream>
#include <string>
#include "dp_const_iterator.h"
#include "dp_exception.h"
#include "dp_iterator.h"
#include "dp_vector.h"
#include "multiple_journey.h"

multiple_journey::multiple_journey(const std::string & a_name)
{
	if (a_name == "")
	{
		throw dp_exception("multiple_journey::multiple_journey(const std::string & a_name) -- a_name == \"\".");
	}
	my_name = a_name;
}

void multiple_journey::output(std::ostream & a_ostream) const
{
	a_ostream << "multiple journey to " << my_name << ":\n";
	dp_const_iterator<journey *> * a_iterator = get_journeys();
	while (!a_iterator->is_done())
	{
		journey * a_journey = a_iterator->get_current();
		if (a_journey == nullptr)
		{
			throw dp_exception("void multiple_journey::output(std::ostream & a_ostream) const -- a_journey == nullptr.");
		}
		a_journey->output(a_ostream);
		a_iterator->advance();
	}
}

void multiple_journey::add_journey(journey * a_journey)
{
	if (a_journey == nullptr)
	{
		throw dp_exception("void multiple_journey::add_journey(journey * a_journey) -- a_journey == nullptr.");
	}
	my_journeys.add(a_journey);
}


dp_const_iterator<journey *> * multiple_journey::get_journeys() const
{
	return my_journeys.make_iterator();
}

dp_iterator<journey *> * multiple_journey::get_journeys()
{
	return my_journeys.make_iterator();
}

File: dp_labs_and_solutions/original/dp_labs_and_solutions/dp_java_labs_and_solutions/singleton/flyweight/lab/MultipleJourney.java
// MultipleJourney.java

import java.io.*;

public class MultipleJourney extends Journey
{
	public MultipleJourney(String aName)
	{
		if (aName == null || aName.equals(""))
		{
			throw new DpException("MultipleJourney.MultipleJourney(String aName) -- aName == null || aName.equals(\"\").");
		}
		myName     = aName;
		myJourneys = new DpVector<Journey>();
	}

	public void output(PrintWriter aPrintWriter)
	{
		if (aPrintWriter == null)
		{
			throw new DpException("void MultipleJourney.output(PrintWriter aPrintWriter) -- aPrintWriter == null.");
		}
        
		aPrintWriter.println("multiple journey to " + myName + ":");
		DpIterator<Journey> aIterator = getJourneys();
		while (!aIterator.isDone())
		{
			Journey aJourney = aIterator.getCurrent();
			if (aJourney == null)
			{
				throw new DpException("void MultipleJourney.output(PrintWriter aPrintWriter) -- aJourney == null.");
			}
			aJourney.output(aPrintWriter);
			aIterator.advance();
		}
	}
    
	public void addJourney(Journey aJourney)
	{
		if (aJourney == null)
		{
			throw new DpException("void MultipleJourney.add(Journey aJourney) -- aJourney == null.");
		}
		myJourneys.add(aJourney);
	}

	public DpIterator<Journey> getJourneys()
	{
		return myJourneys.makeIterator();
	}

	private String            myName;
	private DpVector<Journey> myJourneys;
}
